### ISSUE TITLE
istioctl --set flag should follow Helm syntax

### ISSUE THREAD

**Issue by ostromart** [istio/istio#21909]
There are reports of istioctl --set syntax not exactly following helm. The only place that's ok is if we're adding some useful new functionality like list selection with [key:value].
It would be great if others could chime in here to identify cases where the syntax is different. 
@howardjohn @costinm @esnible

**Comment by esnible** [comment]
I don't know what the differences are.

https://helm.sh/docs/helm/helm_install/ shows that `helm install --set` "can specify multiple or separate values with commas: key1=val1,key2=val2".  We formerly allowed it, but I removed it because https://github.com/istio/istio/issues/20381 and difficulties with meaningful commas e.g `--set values.global.proxy.includeIPRanges="172.30.0.0/16,172.21.0.0/16"`

Helm's syntax is discussed at https://helm.sh/docs/intro/using_helm/

- Lists can be expressed by enclosing values in { and }. For example, `--set name={a, b, c}`
- Array index syntax e.g. `--set servers[0].port=80` **WE HAVE**
- Backslash escaping e.g. `--set name=value1\,value2` and `--set nodeSelector."kubernetes\.io/role"=master` **We have this for dots, not sure about commas.  Plus, the Bash shell gets involved so it is hard to remember what brings what.**

**Comment by dhermes** [comment]
It's unclear, how one even sets a value that happens to occur as part of a list, e.g. `--set 'components.ingressGateways[0].enabled=false'` or `--set 'components.ingressGateways=[]` both fail (in 1.5.0) with

```
Error: failed to generate tree from the set overlay, error: bad path=value: components.ingressGateways[0].enabled=false
...
Error: failed to generate tree from the set overlay, error: bad path=value: components.ingressGateways=[]
```

**Comment by esnible** [comment]
@dhermes `istioctl manifest generate --set components.ingressGateways[0].enabled=false` works with master (the branch that will become Istio 1.6).

It does not work with the release-1.5 branch.  It is possible some of the PRs for lists failed to merge properly.  I will investigate.

**Comment by dhermes** [comment]
Great to hear, thanks!

**Comment by ostromart** [comment]
@esnible I couldn't tell from your comment if we support lists?

**Comment by esnible** [comment]
@ostromart The PR for supporting lists did not make it into 1.5.0 but it merged and will be in 1.5.1 and master.

In 1.5.0 we support `--set` overrides to EXISTING lists, but the PR that merged today for release-1.5 and last month for master lets create and modify lists that aren't in the original profile.

**Comment by ostromart** [comment]
List working as intended now, closing.
